fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * This indexer supports mutual concurrent multi-target indexing by multiple processes or hosts. To do it, each indexer
+ * must be called with the exact same arguments.
+ * The problem that we try to solve:
+ *    Sometimes, building indexes for a store with many records takes too long. To speed up indexing, we wish to perform
+ *    the task in parallel by multiple entities - threads, processes, or hosts.
+ * The problem that we do not try to solve:
+ *    This solution is not optimized for small stores, or for short-lived processes. We can safely assume that the indexing,
+ *    in each process, takes long enough to consider Order(1) metadata-operations as neglectable.
+ * For flexibility, we wish to be able to build index(s) by multiple entities. Whilst:
+ *   - Be consistent - each process can bring the indexing process to success.
+ *   - Be efficient - all the processes should participate in the built until success.
+ *   - Let the user use any arbitrary number of processes, yet keep said efficiency and consistency.
+ *   - Assume no locality - the only interaction between the processes is the fdb database itself.
+ *   - Be able to add indexing processes on the flight.
+ *   - Survive halted/crashed processes while keeping consistency, and with minimal harm to the efficiency.
+ *   - Avoid exposing the implementation details to the caller.
+ *   - Allow a seamless continuation if the indexes were partly built in this mode.
+ * The solution:
+ *   Init:
+ *   - Divide the records space to fragments. Assume that other processes get the same boundaries, but do not break if they do not.
+ *   - Get a random step and a starting point - the step is a prime number that is not a divisor of the fragments count. This
+ *     is designed to provide each process a unique iteration order over the fragments, but will not break if it is not unique.
+ *   Build:
+ *   - One iteration in FULL mode - only build fragments that are fully unbuilt (for efficiency).
+ *   - One iteration in ANY mode - build any fragment that has a missing range.
+ *   - Stop when the indexes are readable, convert to readable if there are no more missing ranges to build.
+ */
+public class IndexingMutuallyByRecords extends IndexingBase {
+    private IndexBuildProto.IndexBuildIndexingStamp myIndexingTypeStamp = null;
+    @Nonnull
+    private static final Logger LOGGER = LoggerFactory.getLogger(IndexingMutuallyByRecords.class);
+
+    List<Tuple> fragmentBoundaries;
+    int fragmentNum;
+    int fragmentStep;
+    int fragmentFirst;
+    int fragmentCurrent;
+    FragmentIterationType fragmentIterationType;
+    Tuple fragmentLastRange;
+    int loopProtectionCounter;
+    String loopProtectionToken = "";
+
+    enum FragmentIterationType {
+        FULL, ANY, RECOVER
```

Comment:
Perhaps worth some javadoc on this enum for what the different iteration types mean

Reply:
It is described at the module's header. 
Added (regular) comments to clarify.